iree/compiler/Conversion/Common/Transforms.cpp

```diff
@@ -261,6 +261,59 @@ class SetWorkgroupSizePattern
 };
 }  // namespace
 
+/// Given the tile sizes to use add a region to the entry point operation that
+/// describes the maximum number of workgroups for a given workload. For now it
+/// is computed as (workload + tilesize - 1) / tilesize along each dimension and
+/// restricted to be 3D dimensional.
+static LogicalResult initNumWorkgroupsRegion(OpBuilder &builder, FuncOp funcOp,
+                                             ArrayRef<int64_t> tileSizes) {
+  auto targetOp =
+      funcOp.getOperation()->getParentOfType<IREE::HAL::ExecutableTargetOp>();
+  IREE::HAL::ExecutableEntryPointOp entryPointOp = nullptr;
+  for (auto op : targetOp.getOps<IREE::HAL::ExecutableEntryPointOp>()) {
+    if (op.sym_name() == funcOp.getName()) {
+      entryPointOp = op;
+      break;
+    }
+  }
+  if (!entryPointOp)
+    return funcOp.emitOpError("unable to find corresponding entry point op");
+  if (entryPointOp.getBody())
+    return entryPointOp.emitOpError("cannot override workgroup_count_region");
+  Location loc = entryPointOp.getLoc();
+
+  OpBuilder::InsertionGuard guard(builder);
+  // Create the cloned operation but with a single region.
+  builder.setInsertionPoint(entryPointOp);
+  auto clonedOp = builder.create<IREE::HAL::ExecutableEntryPointOp>(
+      loc, entryPointOp.sym_nameAttr(), entryPointOp.ordinalAttr(),
+      entryPointOp.interfaceAttr(), entryPointOp.signatureAttr(),
+      entryPointOp.workgroup_sizeAttr(), 1);
+  Region *region = clonedOp.getBody();
+  Block *entryBlock = new Block();
+  region->push_back(entryBlock);
+  builder.setInsertionPointToStart(entryBlock);
+  // Add 3 index arguments for the workload.
+  auto indexType = builder.getIndexType();
+  SmallVector<BlockArgument, 4> workload = llvm::to_vector<4>(
+      entryBlock->addArguments({indexType, indexType, indexType}));
+  // Make the number of workgroups workload / tile size.
+  SmallVector<Value, 4> returnValues;
+  Value one = builder.create<ConstantIndexOp>(loc, 1);
+  assert(tileSizes.size() <= 3 &&
+         "expected only three tile size values for num workgroups computation");
+  for (auto ts : llvm::enumerate(llvm::reverse(tileSizes))) {
+    Value tsVal = builder.create<ConstantIndexOp>(loc, ts.value());
```

Comment:
Can you create an `affine.apply` here?
I have lifted prehistoric usage restrictions re symbols checks and promotions.
It will also be easier to match later.

Bonus points to look up other places in IREE that should use `affine.apply`.

Reply:
ok sounds good. The last time I sued `affine.apply` I ran into the issue you fixed later. Let me try and update.